### PR TITLE
[Form] Improve invalid messages for form types

### PR DESCRIPTION
**Pull request by hiddewie**
| Q             | A
| ------------- | ---
| Bug fix?      |  no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #5946   
| License       | MIT

Improvement of #14106. Added `invalid_message` fields to all form types. This gives better errors to users instead of the default *Invalid value* error.

### Backwards compatibility

After discussion about BC, the following additions have been made:
- An framework option `framework.form.improved_validation_messages` (default false) for enabling improved validation messages globally.
- A form validation extension option `improved_validation_messages` which enables improved validation messages for that type.
- An added parameter for the `ValidatorExtension` and `FormTypeValidatorExtension` (default false) to enable/disable the improved validation messages for that extension.
- All the form types now check for the enabled opt-in configuration for improved validation messages. Otherwise, the previous default is used which should be the original message.

### Upgrade path

- These changes should be backwards compatible.
- For Symfony 5.0, the improved validation messages configuration in the framework and form should be `true` (enabled) by default.
- The options can be deprecated entirely in Symfony 5.1, to be removed in Symfony 6.0.

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/MoneyType.php
Comment (fabpot) [inline src/Symfony/Component/Form/Extension/Core/Type/MoneyType.php]: money amount?

**Review by fabpot — CHANGES_REQUESTED** [review]
Looks good to me. We should add all these new error messages in translations as well.

**Comment by hiddewie** [comment]
@fabpot Thank you for the review. I've added your comment for the translation for the money type. As for the translations, do you mean all the files in `src\Symfony\Component\Form\Resources\translations\validators.*.xlf`? Of course I can add the fields there, but I cannot do translations in all those different languages (except Dutch and maybe German).

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: `time zone`

File: src/Symfony/Component/Form/Extension/Core/Type/TextareaType.php
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TextareaType.php]: i think we can keep the default value here no? same for `text` and other generic types

Comment (hiddewie) [reply src/Symfony/Component/Form/Extension/Core/Type/TextareaType.php]: Yes, I've removed a few custom translations from generic types.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php
Comment (hiddewie) [reply src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: I've updated the message, although the `Type` is called `TimezoneType` which may be better called `TimeZoneType`

**Comment by hiddewie** [comment]
@fabian I've added the translations to the English file. However, I am not sure what the correct precedure is for all the other languages (of which I can only translate 2 myself).

**Comment by xabbuh** [comment]
I am not sure that it is a good idea to change the default value for this option for all built-in form types. This means that custom translations in project wouldn't be working anymore after updating to a Symfony version containing this feature. That's not a good DX IMO.

**Comment by hiddewie** [comment]
@xabbuh You are right, these changes could be seen as backwards incompatible. What do you propose: close the PR or leave it open until there is a major new version?

**Comment by fabpot** [comment]
I didn't thought about the BC issue here. Looks like a deal breaker :( Let's close then. Sorry @hiddewie

**Comment by ogizanagi** [comment]
Perhaps we could reopen and find a proper upgrade path?

For instance, register a `better_invalid_messages` or [INSERT_BETTER_NAME] option from `FormTypeValidatorExtension`. Then if, if `better_invalid_messages` is `false`, trigger a deprecation.
A `framework.form.better_invalid_messages` framework option can be used to configure the default value used by `FormTypeValidatorExtension` globally, so any new project can be free of deprecations.

Then, in 5.1, deprecate `better_invalid_messages` and `framework.form.better_invalid_messages` to be removed in 6.0 as it won't have any effect anymore.

This seems a lot, but I find this PR valuable. Actually, IMHO, the BC break here is sensible.

**Comment by hiddewie** [comment]
@ogizanagi That would be possible! Would an option name of `improved_validation_messages` work?

What is the proper precedure for deprecating something in a far future release? Just open a new issue after this PR is merged (if it is merged at all)?

**Comment by ogizanagi** [comment]
> What is the proper procedure for deprecating something in a far future release? Just open a new issue after this PR is merged (if it is merged at all)?

Actually I don't think we have any specific workflow for this.
An issue could be great, but can be easily forgotten.
We could open a note on a project in https://github.com/symfony/symfony/projects in order to not forget this, but this kind of situation doesn't happen really often.
But I'm confident anyone doing the PR to remove the BC layer in 5.0 will think about doing one to deprecate it in 5.1 :)

**Comment by hiddewie** [comment]
@ogizanagi I've made some changes. Review is very welcome. The PR has grown considerably in code and also in complexity, but I hope everything works out. 

Also see the updated description of the PR for some notes on the upgrade path. Mostly what you described.

File: src/Symfony/Component/Form/Extension/Core/Type/BirthdayType.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/BirthdayType.php]: This condition could be simplified everywhere: 

```diff
- isset($options['improved_validation_messages']) && true === $options['improved_validation_messages']
+ $options['improved_validation_messages'] ?? false
```

Also the ternary should be inlined according to codebase code style

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: Could be simplified using a early return, avoiding the else clause :)

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: Should we also add allowed types to the resolver for this option?

File: src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php]: I believe the proper cs here is:

```diff
- function(Options $options, $previousValue) {
+ function (Options $options, $previousValue) {
```

There are a lot more occurrences. I thought fabbot would have smelt these.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I don't know if we can improve this sentence, but it will not only be `true` by default in 5.0; it'll always be true and previous messages dropped.
Perhaps:

> Setting the "framework.form.improved_validation_messages" to \'false\' has been deprecated and will have no effect as of Symfony 5.0.

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: Missing `bool` typehint :)

File: src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php]: Missing `bool` typehint

File: src/Symfony/Component/Form/Resources/translations/validators.en.xlf
Comment (ogizanagi) [inline src/Symfony/Component/Form/Resources/translations/validators.en.xlf]: 32

**Review by ogizanagi — COMMENTED** [review]
Great work @hiddewie
Note we could set this to true by default for new projects by adding a form recipe.

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (hiddewie) [reply src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: Added

File: src/Symfony/Component/Form/Extension/Core/Type/BirthdayType.php
Comment (hiddewie) [reply src/Symfony/Component/Form/Extension/Core/Type/BirthdayType.php]: Yeah, much cleaner.

**Comment by hiddewie** [comment]
@ogizanagi I've processed your comments. Thanks for the input.

- I will look into updating the receipe.
- I also want to write tests for at least some form types in the Core extension and using the Type Validator Extension which check the old and the new error messages.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (ogizanagi) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Indeed. I'd remove it.

File: src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php
Comment (ogizanagi) [reply src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php]: That'd be less flexible (as it'll only be global) if you want to update part of your app step by step...but regarding the number of translations to update, that might be an option to consider.

Comment (ro0NL) [reply src/Symfony/Component/Form/Extension/Validator/ValidatorExtension.php]: introducing a new `improved_validation_messages` form type option, we have to deprecate again in 5.0 and remove in 6.0, is not worth it IMHO

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (hiddewie) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Done

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (hiddewie) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: How do you inject it if not setting it as a container parameter? I'm comparing with `validator.translation_domain` and `form.type_extension.csrf.enabled` for example.

**Comment by hiddewie** [comment]
@ro0NL @ogizanagi Processed some comments.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (ogizanagi) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: By defining an empty argument in the `form.type_extension.form.validator` xml definition & getting the definition back in the extension to replace the second argument.

Comment (hiddewie) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: :+1: Done

**Review by HeahDude — CHANGES_REQUESTED** [review]
Thank you for improving this in a BC way!
It would be nice to have other reviews to be sure we cannot improve some wording more before breaking BC later. @weaverryan @javiereguiluz would you mind having a look at this one?

**Comment by hiddewie** [comment]
@HeahDude I've processed your comments. Didn't think of the types which cannot be invalid.

**Comment by HeahDude** [comment]
Thank you @hiddewie!

I've reviewed it again and it looks good to me. But we need to adjust some things now that #28731 as been merged. Because the option did not exist in core before, only added by the validator extension (https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php#L59).

So here's what I spotted so far:

- the option can be handled here now https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/EventListener/TransformationFailureListener.php#L54.
- the check here need to be changed https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/TransformationFailureExtension.php#L33
 Maybe we should rely on the `constraints` option only.

ping @xabbuh what do you think?

**Comment by hiddewie** [comment]
@HeahDude Thanks. Processed comment.

I'm unsure if I should remove the same code from https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Validator/Constraints/FormValidator.php#L117 (I don't think so since any forms with the `constraints` option are excluded from the new extension).

**Comment by fabpot** [comment]
@hiddewie For this pull request to be merged, you need to rebase it (to remove the merge commits).

**Comment by hiddewie** [comment]
@fabpot Done.

File: src/Symfony/Component/Form/Extension/Core/Type/FormType.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/FormType.php]: You must define the `invalid_message_parameters` option here now to be able to use it in the core extension. It also means we're missing a test case for this support.

File: src/Symfony/Component/Form/Extension/Core/Type/FormType.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/FormType.php]: We don't need the previous value here, so this can be simplified to:
```php
'invalid_message' => 'The form is invalid',
```

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: This must be change for BC:
```php
'invalid_message' => function (Options $options, ?string $previousValue) {
    return $options['legacy_error_messages'] ? 'This value is not valid.' : $previousValue;
},
```
We need a test case for this. Could you please add a test case in the BaseTypeTest to check the default value?

File: src/Symfony/Component/Form/Resources/translations/validators.en.xlf
Comment (HeahDude) [inline src/Symfony/Component/Form/Resources/translations/validators.en.xlf]: Where is 32?

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/BirthdayTypeTest.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/BirthdayTypeTest.php]: Must be reverted.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php]: Can be reverted and merged in 3.4 to avoid merging conflicts between branches.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php]: Can be reverted.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/FormTypeTest.php]: Should be reverted too.

**Review by HeahDude — CHANGES_REQUESTED** [review]
Thank you! another round of review :)

**Comment by hiddewie** [comment]
@HeahDude Thanks for the review. I've added a load of tests for the custom validation messages, including the legacy variant.

I also noticed that the logic for legacy / non-legacy was flipped in the form types. Has been corrected.

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: `OptionsResolver` should be `Options`

File: src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php]: I would use double quotes here too for enclosing option names and values (same as above)

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by xabbuh** [comment]
Thank you very much for your pull request!

As part of the Symfony EU funded hackathon (https://symfony.com/blog/the-symfony-and-api-platform-hackathon-is-coming), we were able to assemble most of the core team and big contributors in one place. Our goal is to finish as many open issues and PRs as possible.

*Your commits will not be lost and you will therefore get credit for your work*. All that will happen is that your commits will be moved to a new PR (see #30931) where all remaining concerns will be addressed.

Without your work this would not have been possible. So thank you once again!

**Comment by hiddewie** [comment]
No problem, thanks for the work!

**Pull request closed without merging**
